megablocks/layers/glu.py

```diff
@@ -0,0 +1,75 @@
+from megablocks.layers import common
+from megablocks.layers import gelu
+from megablocks.layers.mlp import SparseMLP, create_dmoe_expert_weights
+from megablocks.layers import mpu
+from megablocks.layers.arguments import Arguments, InitFn
+from megablocks import grouped_gemm_util as gg
+import stk
+import torch
+import torch.nn.functional as F
+
+
+class SparseGLU(SparseMLP):
+
+    def __init__(self, args : Arguments):
+        super().__init__(args)
+        num_rows_per_rank = (
+            (mpu.experts_per_rank(args) * mpu.features_per_rank(args)) //
+            mpu.get_weight_parallel_world_size(args)
+        )
+
+        self.v1 = torch.nn.Parameter(torch.empty(
+            num_rows_per_rank,
+            args.hidden_size,
+            device=args.device,
+            dtype=common.dtype(args)))
+        with torch.no_grad():
+            self.v1.copy_(create_dmoe_expert_weights(
+                args, args.moe_num_experts, args.ffn_hidden_size,
+                args.hidden_size, args.init_method))
+
+        should_set_attribute = (
+            args.moe_expert_model_parallelism or args.moe_weight_parallelism)
+
+        mpu.set_expert_model_parallel_attributes(
+            self.v1, should_set_attribute)
+
+    def forward(self, x, topo):
+        w1, v1, w2 = (self.scale_grad(self.w1), self.scale_grad(self.v1), self.scale_grad(self.w2))
+        if self.args.moe_weight_parallelism:
+            raise NotImplementedError("Weight parallelism not yet supported with GLU.")
+        elif self.args.memory_optimized_mlp:
+            raise NotImplementedError("Memory optimized implementation not yet supported with GLU.")
+
+        # Compute the GLU.
+        x1 = stk.ops.sdd(x, w1.t(), topo)
+        x2 = stk.ops.sdd(x, v1.t(), topo)
+
+        x1._data = gelu.gelu(x1)._data * x2._data
+
+        return stk.ops.dsd(x1, w2)
+    
+class GroupedGLU(SparseGLU):
+    def forward(self, x, tokens_per_expert):
+        batch_sizes = tokens_per_expert.cpu().to(torch.long)
+        w1, v1, w2 = (self.scale_grad(self.w1), self.scale_grad(self.v1), self.scale_grad(self.w2))
+
+        # Re-shape the weights for the grouped GEMMs.
```

Comment:
Maybe private helper in the base class for this?

Reply:
For reshaping the weights, this one is a bit weird with my current inheritance structure, where I have `GroupedGLU` derived from `SparseGLU`, not from `GroupedMLP` (which is where it would make sense to add a helper). Thoughts?

Reply:
Oh huh. I'd inherit from `GroupedMLP`, maybe? And then these would go away anyways?

Reply:
I structured it that way so I could avoid having to remake the `__init__` function for setting up the v1 matrix, but I can refactor that if inheriting from `GroupedMLP` makes more sense.

Reply:
tho then I would have basically the same `__init__` block twice...
``` 
class GroupedGLU(GroupedMLP):
...
def __init__(self, args : Arguments):
        super().__init__(args)
        self.v1 = torch.nn.Parameter(torch.empty(
            self._num_rows_per_rank,
            args.hidden_size,
            device=args.device,
            dtype=common.dtype(args)))
        with torch.no_grad():
            self.v1.copy_(create_dmoe_expert_weights(
                args, args.moe_num_experts, args.ffn_hidden_size,
                args.hidden_size, args.init_method))
```

Reply:
If you want, you could make a helper to create `v1` for both. That's not necessary if you don't want to, though.